File: 05-binary_search/00153-Find_Minimum_in_Rotated_Sorted_Array/00153-find_minimum_in_rotated_sorted_array.cpp
// 153: Find Minimum in Rotated Sorted Array
// https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/


#include <iostream>
#include <vector>

using namespace std;


class Solution {
public:
    // SOLUTION
    int findMin(vector<int>& nums) {
        int l = 0;
        int r = nums.size() - 1;
        
        while (l<r && nums[l]>nums[r]) {
            int m = l + (r-l) / 2;

            if (nums[m] > nums[r])
                l = m + 1;
            else
                r = m;
        }

        return nums[l];
    }
};


int main() {
    Solution o;

    // INPUT
    vector<int> nums = {3,4,5,1,2};

    // OUTPUT
    auto result = o.findMin(nums);
    cout << result << endl;

    return 0;
}

File: 05-binary_search/00153-Find_Minimum_in_Rotated_Sorted_Array/00153-find_minimum_in_rotated_sorted_array.go
// 153: Find Minimum in Rotated Sorted Array
// https://leetcode.com/problems/find-minimum-in-rotated-sorted-array/

package main

import "fmt"

// SOLUTION
func findMin(nums []int) int {
    l := 0
    r := len(nums) - 1
    
    for (l<r && nums[l]>nums[r]) {
        m := l + (r-l) / 2

        if (nums[m] > nums[r]) {
            l = m + 1
        } else {
            r = m
        }
    }

    return nums[l]
}


func main() {
    // INPUT
    nums := []int{3,4,5,1,2}

    // OUTPUT
    result := findMin(nums)
    fmt.Println(result)
}